### ISSUE TITLE
cmake error on a clean build

### ISSUE THREAD

**Issue by wyq977** [bsamseth/cpp-project#21]
Thx a lot for sharing the repo! I have been trying to find a repo template for cpp with support for CI and your repo is a great template for my project.

However, I was unable to run `cmake .. -DCMAKE_BUILD_TYPE=coverage` after following the steps. 

The error is as follow 
```
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 
-- IPO is supported
CMake Error at CMakeLists.txt:58 (target_link_libraries):
  Object library target "example" may not link to anything.


-- Configuring incomplete, errors occurred!
See also "/workspaces/PoissonDiscSampling/build/CMakeFiles/CMakeOutput.log".
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# mkdir build && cd build && cmake .. -DCMAKE_BUILD_TYPE=coverage^C
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# cd ../^C
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# rm -rf ./*
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# mkdir build && cd build && cmake .. -DCMAKE_BUILD_TYPE=coverage -DENABLE_DOCTESTS=OFF
CMake Error: The source directory "/workspaces/PoissonDiscSampling/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build/build# cd ..
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# rm -rf ./*
root@c067bbd707d0:/workspaces/PoissonDiscSampling/build# cmake .. -DCMAKE_BUILD_TYPE=coverage -DENABLE_DOCTESTS=OFF
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 
-- IPO is supported
CMake Error at CMakeLists.txt:58 (target_link_libraries):
  Object library target "engine" may not link to anything.


-- Configuring incomplete, errors occurred!
```
Thought it might be related to the git submodule status, I try using `git submodule add ` to manually add doctest but still no luck.

__code to reproduce__ (cmake version 3.10.2):
```
git clone https://github.com/bsamseth/cpp-project.git
cd cpp-project
sh setup.sh
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=coverage
```

Github Action would also return failure during build step, see [here](https://github.com/wyq977/cpp-example/runs/526486319?check_suite_focus=true)

**Comment by wyq977** [comment]
After looking through the github action CI, I found out that the git submodule failed to clone into `external/doctest` where it later led to errors in cmake. As indicated by the messages in logging:

Successful run:
```
Run auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  git submodule sync --recursive
  git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
Submodule 'external/doctest' (https://github.com/onqtam/doctest.git) registered for path 'external/doctest'
Cloning into '/home/runner/work/cpp-project/cpp-project/external/doctest'...
Submodule path 'external/doctest': checked out '503de03ee2cbbfc58c2063dce36dc96f917f2da3'
```

Failed run (__without git submodule__)
```
Run auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  git submodule sync --recursive
  git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
```

**Comment by bsamseth** [comment]
Hi!

Thanks for reporting this. Submodules in git always cause issues. Still, I realize it is not made as clear as it should that you need submodules for the build to work. I also noticed that the `setup.sh` script could fail under some versions of `sh` (I changed computer, and now it fails) due to me using `bash` syntax. This would cause the submodule stuff to be skipped.

I've made some changes to both the README and the setup script to improve this (#22). It seems to do the trick, at least on my machine.

In summary: When cloning, use `git clone --recurse-submodules <url>`. If you already cloned without, it _should_ be enough to do `git submodule update --init --recursive`. But the safest is to clone the "_right_" way.

Hope it helps!

**Comment by wyq977** [comment]
Thanks a lot for the tips, forgot to mention the `setup.sh` but great that you fixed it!

I would suggested made some twists in the github action as checkout support recursive when cloning

see https://github.com/actions/checkout/issues/34

So the modified workflow would start with:
```
    - uses: actions/checkout@master
      with:
        submodules: true
```

It works on other example but I have no luck with that since they just merged it into master branch 4 days ago. But `textbook/git-checkout-submodule-action@master` should works in theory

**Comment by bsamseth** [comment]
Thanks for the input about the simplified action - I don't remember where I got the code I use now, but in any case this is cleaner. It works fine. As soon as all the CI checks complete I'll merge the PR and `master` should be good to go.
